system_metrics/system_metrics_client.go

```diff
@@ -174,6 +183,179 @@ func (s *SystemMetricsClient) AddSystemMetrics(ctx context.Context, id string, d
 	return err
 }
 
+// SystemMetricsWriteCloser is a wrapper around a stream of system metrics data
+// that implements buffering with timed flushes and the WriteCloser interface.
+type SystemMetricsWriteCloser struct {
+	mu            sync.Mutex
+	ctx           context.Context
+	cancel        context.CancelFunc
+	catcher       grip.Catcher
+	id            string
+	stream        internal.CedarSystemMetrics_StreamSystemMetricsClient
+	buffer        []byte
+	maxBufferSize int
+	lastFlush     time.Time
+	timer         *time.Timer
+	flushInterval time.Duration
+	closed        bool
+}
+
+// Write adds provided data to the buffer, flushing it if it exceeds its max
+// size. Write will fail if the writer was previously closed or if a timed flush
+// encountered an error.
+func (s *SystemMetricsWriteCloser) Write(data []byte) (int, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.catcher.HasErrors() {
+		return 0, errors.Wrapf(s.catcher.Resolve(), "writer already closed due to error")
+	}
+	if s.closed {
+		return 0, errors.New("writer already closed")
+	}
+	if len(data) == 0 {
+		return 0, errors.New("must provide data to write")
+	}
+
+	s.buffer = append(s.buffer, data...)
+	if len(s.buffer) > s.maxBufferSize {
+		if err := s.flush(); err != nil {
+			s.catcher.Add(err)
+			s.catcher.Add(s.close())
+			return 0, errors.Wrapf(s.catcher.Resolve(), "problem writing data")
+		}
+	}
+	return len(data), nil
+}
+
+// Close flushes any remaining data in the buffer and closes the stream. If
+// the writer was previously closed, this returns an error.
+func (s *SystemMetricsWriteCloser) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.catcher.HasErrors() {
+		return errors.Wrapf(s.catcher.Resolve(), "writer already closed due to error")
+	}
+	if s.closed {
+		return errors.New("writer already closed")
+	}
+
+	s.catcher.Add(s.flush())
+	s.catcher.Add(s.close())
+	return s.catcher.Resolve()
+}
+
+func (s *SystemMetricsWriteCloser) flush() error {
+	data := &internal.SystemMetricsData{
+		Id:   s.id,
+		Data: s.buffer,
+	}
+
+	if err := s.stream.Send(data); err != nil {
+		return errors.Wrapf(err, "problem sending data for id %s", s.id)
+	}
+
+	s.buffer = []byte{}
+	s.lastFlush = time.Now()
+	return nil
+}
+
+func (s *SystemMetricsWriteCloser) close() error {
+	s.cancel()
+	s.closed = true
+	_, err := s.stream.CloseAndRecv()
+	return err
+}
+
+func (s *SystemMetricsWriteCloser) timedFlush() {
+	defer func() { s.catcher.Add(recovery.HandlePanicWithError(recover(), nil, "op")) }()
```

Comment:
Catching this is good, but can we also log the error with `grip.Error()` as soon as it occurs?